rpm/sqlite/sqlite.go

```diff
@@ -0,0 +1,169 @@
+// Package sqlite extracts RPM package information from SQLite databases.
+package sqlite
+
+import (
+	"context"
+	"database/sql"
+	_ "embed" // embed a sql statement
+	"fmt"
+	"net/url"
+	"runtime"
+
+	_ "modernc.org/sqlite" // register the sqlite driver
+)
+
+// RPMDB is a handle to a SQLite RPM database.
+type RPMDB struct {
+	db *sql.DB
+}
+
+// Open opens the named SQLite database and interprets it as an RPM
+// database.
+//
+// Must be a file on-disk. This is a limitation of the underlying SQLite
+// library.
+//
+// The returned RPMDB struct must have its Close method called, or the
+// process may panic.
+func Open(f string) (*RPMDB, error) {
+	u := url.URL{
+		Scheme: `file`,
+		Opaque: f,
+		RawQuery: url.Values{
+			"_pragma": {
+				"foreign_keys(1)",
+				"query_only(1)",
+			},
+		}.Encode(),
+	}
+	db, err := sql.Open(`sqlite`, u.String())
+	if err != nil {
+		return nil, err
+	}
+	rdb := RPMDB{db: db}
+	_, file, line, _ := runtime.Caller(1)
+	runtime.SetFinalizer(&rdb, func(rdb *RPMDB) {
+		panic(fmt.Sprintf("%s:%d: RPM db not closed", file, line))
+	})
+	return &rdb, nil
+}
+
+// Close releases held resources.
+//
+// This must be called when the RPMDB is no longer needed, or the
+// process may panic.
+func (db *RPMDB) Close() error {
+	runtime.SetFinalizer(db, nil)
+	return db.db.Close()
+}
+
+// Packages loads all package headers in the RPMDB and reports [Info]s
+// for each.
+func (db *RPMDB) Packages(ctx context.Context) ([]Info, error) {
+	hs, err := db.loadHeaders(ctx)
+	if err != nil {
+		return nil, err
+	}
+	out := make([]Info, len(hs.key))
+	for i, k := range hs.key {
+		p := &out[i]
+		h := hs.header[k]
+		for i := range h.Infos {
+			e := &h.Infos[i]
+			if _, ok := wantTags[e.Tag]; !ok {
+				continue
```

Comment:
It would if there's somehow a header with no entries that we're interested in, which given that we're interested in the main 5 tags (Name, Epoch, Version, Release, Architecture) would be exceedingly weird. I could add a belt-and-suspenders check if you'd like, though.